exmg-paper-token-input.ts

```diff
@@ -0,0 +1,546 @@
+   * look up the property of a selectable element. Camel cased values are
+   * *not* converted to hyphenated values for attribute lookup. It's
+   * recommended that you provide the hyphenated form of the name so that
+   * selection works in both cases. (Use `attr-or-property-name` instead of
+   * `attrOrPropertyName`.)
+   */
+  @property({type: String, attribute: 'attr-for-selected'})
+  public attrForSelected?: string;
+
+  /**
+   * By default the textContent of the paper-item/paper-icon-item or paper-item-body
+   * will be used for display in badge after selection. In case of icon and body
+   * you probably want an alternative. The selector can be used to be a bit more
+   * specific on which element can be used for display purposes.
+   */
+  @property({type: String, attribute: 'selected-item-selector'})
+  public selectedItemSelector?: string;
+
+  /**
+   * Gets or sets the selected elements.
+   */
+  @property({type: Array, attribute: 'selected-values'})
+  public selectedValues: SelectedValue[] = []; // @todo migrate notify: true
+
+  /**
+   * The label for this input.
+   */
+  @property({type: String})
+  public label?: string;
+
+  /**
+   * Set to true to auto-validate the input value.
+   */
+  @property({type: Boolean, attribute: 'auto-validate'})
+  public autoValidate: boolean = false;
+
+  @property({type: Boolean})
+  public autofocus: boolean = false;
+
+  /**
+   * Set to true to disable this input.
+   */
+  @property({type: Boolean})
+  public disabled: boolean = false;
+
+  /**
+   * The error message to display when the input is invalid.
+   */
+  @property({type: String, attribute: 'error-message'})
+  public errorMessage?: string;
+
+  /**
+   * alwaysFloatLabel
+   */
+  @property({type: Boolean})
+  public alwaysFloatLabel: boolean = false;
+
+  /**
+   * Maximum number of tokens allowed in value
+   */
+  @property({type: Number, attribute: 'max-tokens'})
+  public maxTokens?: number;
+
+  /**
+   * Set to true to mark the input as required. If you're using PaperInputBehavior to
+   * implement your own paper-input-like element, bind this to
+   * the `<input is="iron-input">`'s `required` property.
+   */
+  @property({type: Boolean})
+  public required: boolean = false;
+
+  /**
+   * This field will be bind to the actual input field
+   */
+  @property({type: String})
+  private inputValue: string = '';
+
+  @property({type: Boolean})
+  public invalid: boolean = false;
+
+  @property({type: Boolean})
+  public inputFocused: boolean = false;
+
+  @property({type: Boolean})
+  private opened: boolean = false;
+
+  @query('#listbox')
+  private listBoxNode?: PaperListboxElement;
+
+  @query('#inputValue')
+  private inputValueNode?: HTMLInputElement;
+
+  @query('#inputWidthHelper')
+  private inputWidthHelperNode?: HTMLElement;
+
+  @query('#menu')
+  private menuElement?: PaperMenuButton;
+
+  constructor() {
+    super();
+    this.onIronInputKeyDown = this.onIronInputKeyDown.bind(this);
```

Comment:
You don't need to bind everything to this only such functions which you use as handlers fro manual added listeners